base/src/main/java/uk/ac/ebi/atlas/search/analyticsindex/FullAnalyticsSearchService.java

```diff
@@ -0,0 +1,251 @@
+
+        SolrQuery solrQuery = new SolrQuery();
+        solrQuery.setRows(0)
+                .setFilterQueries(filterQueries)
+                .setQuery(
+                        MessageFormat.format(SELECTED_ASSAY_GROUP_IDS_Q_TEMPLATE, Joiner.on(",").join(assayGroupIds), expressionThreshold))
+                .set("json.facet", ResourceUtils.readPlainTextResource(avgExpressionOverBioentityIdentifiersFacet));
+
+        Stopwatch stopwatch1 = Stopwatch.createStarted();
+        LOGGER.debug(
+                "Searching for genes in {} {} across {} assay groups...",
+                experimentAccession, expressionThreshold, assayGroupIds.size());
+
+        TreeMultimap<Integer, Pair<String, Double>> genesBySpecificity =
+                parseAverageExpressionOverSelectedAssayGroups(
+                        fetchResponseAsString("http://localhost:8983/solr/analytics/query", solrQuery), resultSize);
+
+        LOGGER.debug("Finished in {} ms", stopwatch1.elapsed(TimeUnit.MILLISECONDS));
+
+        Map<String, Double> avgExpressionByGeneId = genesBySpecificity.keySet().stream()
+                .flatMap(specificity -> genesBySpecificity.get(specificity).stream())
+                .collect(Collectors.toMap(Pair::getLeft, Pair::getRight));
+
+
+        String nonSelectedQuery = avgExpressionByGeneId.keySet().size() > TERMS_QUERY_SIZE_THRESHOLD ?
+                MessageFormat.format(
+                        NON_SELECTED_ASSAY_GROUP_IDS_Q_TEMPLATE_ALT,
+                        Joiner.on(",").join(assayGroupIds),
+                        expressionThreshold) :
+                MessageFormat.format(
+                        NON_SELECTED_ASSAY_GROUP_IDS_Q_TEMPLATE,
+                        Joiner.on(",").join(assayGroupIds),
+                        Joiner.on(",").join(avgExpressionByGeneId.keySet()),
+                        expressionThreshold);
+
+        solrQuery.setQuery(nonSelectedQuery)
+                .set("json.facet", ResourceUtils.readPlainTextResource(maxExpressionOverBioentityIdentifiersFacet));
+
+        Stopwatch stopwatch2 = Stopwatch.createStarted();
+        LOGGER.debug(
+                "Searching for genes in {} {} across inv({}) assay groups and {} gene IDs...",
+                experimentAccession, expressionThreshold, assayGroupIds.size(), avgExpressionByGeneId.keySet().size());
+
+        Map<String, Double> maxExpressionByGene =
+                parseMaximumExpressionOverNonSelectedAssayGroups(
+                        fetchResponseAsString("http://localhost:8983/solr/analytics/query", solrQuery));
+
+        LOGGER.debug("Finished in {} ms", stopwatch2.elapsed(TimeUnit.MILLISECONDS));
+
+
+        TreeMultimap<Integer, GeneProfileScore> rankedGenesBySpecificity =
+                TreeMultimap.create(
+                        Comparator.naturalOrder(),
+                        GeneProfileScore.COMPARATOR_BY_MAX_EXPRESSION.reversed());
+
+        genesBySpecificity.keySet()
```

Comment:
Maybe the procedural solution?
```
List<String> result = new ArrayList<>();
Iterator<Integer> specificities = genesBySpecificity.keySet().sort().reverse().iterator();
Integer specificity;
while(specificities.hasNext() && result.size() < MAX_SIZE) {
  result.addAll(
   genesBySpecificity.get(specificities.next()).stream()
                                         .map(pair ->
                                                 GeneProfileScore.create(
                                                        pair.getLeft(),
                                                         pair.getRight() /
                                                                 maxExpressionByGene.getOrDefault(pair.getLeft(), 1.0)))
                                         .takeAtMostElements(MAX_SIZE - result.size()) // I forgot the method name 
                                         .collect(Collectors.toList()) // I think you want toList, so that the results are ordered? 
);
}
return result;
```

Reply:
Once we have the specified order, we'll do one more query, and construct a BaselineProfilesList - which you are doing now if I get this correctly?

Reply:
Yes, that’s the idea once we have the 50 (or however many) gene IDs sorted. Not yet done though.